# sciTIP-Seq analysis
Scripts for sciTIP-Seq data processing

Data processing scripts for H3.3 mESC sciTIP-Seq experiment.
The majority of the code is adapted from the original [Bartlett et al. paper](https://rupress.org/jcb/article/220/12/e202103078/212821/High-throughput-single-cell-epigenomic-profiling)

Additional codes: [Bartlett et al. github repo](https://github.com/dbart1807/TIP-seq)

### Steps: 
1. relocalize FASTQ files of each 96-well plate row. 
2. bowtie2 alignment to mm10 genome
3. BAM convertion and deduplications (samtools)
4. T7 deduplications
5. computing read count matrix (bedtools) 

You can run the whole pipeline by one script called _all.R_

Its parameters:

--input (-i): absolute path with the folder where the deindexed fastqs can be found

--workdir (-w): name of existing working directory where all the folders and results will be created

Important folders at Uppmax:

  * Linxuan folder:
  _/proj/snic2020-6-3/LINXUAN/sciTIP_sc_analysis/_

  * Working directory:
  _/proj/snic2020-6-3/SZABOLCS/sciTIP-Seq/utils_
 
 
 # sciTIP-Seq, allele-specific alignment
 
 ### Steps
 
1. bowtie2 alignment to N-masked mm10 genome (_bowtie2_allele_spec.R_)

N-masking was generated by SNPsplit tool:

SNPsplit_genome_preparation --vcf_file mgp_REL2021_snps.vcf.gz --strain CAST_EiJ --reference_genome /proj/snic2020-6-3/SZABOLCS/sciTIP-Seq/data/mm10_reference
output was indexed by bowtie2

[SNPsplit_genome_preparation](https://felixkrueger.github.io/SNPsplit/genome_prep/genome_preparation/)
 
2. BAM convertions and deduplications (samtools) (_bowtie2_allele_spec.R_)
3. SNPsplit (_snp_split.sh_)
	-i: input directory with the processed BAM files
	-w: name of working dir
	-o: directory for SNPsplit outputs
	
	_snp_split.sh_ runs snp_split.R, that contains the main SNPsplit command:
	
	SNPsplit --snp_file VCF_file.vcf processed_bam.bam
	
		*VCF_file is the VCF file with the SNP informations.
		*Processed bam comes from the samtools processing of _bowtie2_allele_spec.R_
	
	[SNPsplit](http://felixkrueger.github.io/SNPsplit/)
  
